schema/validator.go

```diff
@@ -92,7 +99,56 @@ func (v unimplemented) Validate(src io.Reader) error {
 	return fmt.Errorf("%s: unimplemented", v)
 }
 
-func validateManifestDescendants(r io.Reader) error {
+type validateRefMediaFunc func(r io.Reader) error
+
+var vlidateRefMediaMap = map[Validator]validateRefMediaFunc{
+	MediaTypeManifestList: validateManifestListRefObject,
+	MediaTypeManifest:     validateManifestRefObject,
+}
+
+// ValidateRefMedia validates the referenced objects against OCI media type defined in spec.
+// r: the given reader.
+// v: the expected media type.
+func ValidateRefMedia(r io.Reader, v Validator) error {
+	f, ok := vlidateRefMediaMap[v]
+	if ok {
+		if f == nil {
+			return fmt.Errorf("referenced media type validation %q unimplemented", v)
+		}
+
+		buf, err := ioutil.ReadAll(r)
+		if err != nil {
+			return errors.Wrap(err, "unable to read the document file")
+		}
+
+		return f(bytes.NewReader(buf))
+	}
+
+	return nil
+}
+
+func validateManifestListRefObject(r io.Reader) error {
+	header := v1.ManifestList{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
+	if err != nil {
+		return errors.Wrap(err, "manifest list format mismatch")
+	}
+
+	for _, manifest := range header.Manifests {
+		if manifest.MediaType != string(v1.MediaTypeImageManifest) {
+			return fmt.Errorf("manifest %s has an unknown media type: %s", manifest.Digest, manifest.MediaType)
```

Comment:
> I don't think it belongs to this patch.

I agree that we don't want to cover the reserved `data` as part of this PR.  This PR is about two things:

* Broadly, setting a precedent for handling verification for behaviour which images MAY use but which implementations MUST support.
* Narrowly, adding optional validation for per-location media types, which are a special case of the broader issue.

So while I have no desire to expand the narrow target, I want to try and solve it in a way that will cover the broader issue.  I think the `strict` option sets a precedent which will scale well as we extend it to other spec statements.

I'm less clear on how the “pass in an array of validator functions” will scale.  It seems like it might end up with a separate validator for each spec statement?  Or maybe we'll end up with a `ValidateOnlyContainsOCIRequiredSettings` validator?  Or…?  It's not clear to me how we're expected to apply this approach to future spec statements.